### DISCUSSION TITLE
Running recipes as standalone applications

### DISCUSSION THREAD

**Discussion by kjy5** [openrewrite/rewrite#4265] [Q&A] [answered]
Hi, I'm new to Rewrite and I'm interested in using it to parse java files into their LSTs and export them as an XML file.

I understand I can write a recipe to do this, but recipes are designed to be added to a project as a dependency and ran internally as a tool. Instead, I would like to use this recipe like a standalone java application (like a JAR file) and pass it in java source files as CLI arguments.

How can I accomplish this? Thank you!

**Comment by timtebeek** [comment, marked as answer]
Hi @kjy5 . There would be a couple of complications with this approach, so I'm wondering if OpenRewrite is the right choice for you. What would you like to use this for?

One complication is that our parsers require you to set up the classpath correctly, which is no easy feat if you look to our Maven and Gradle plugin handling. Another complication is that serialization of LSTs has explicitly not been open sourced, as we view that as unique what we do at Moderne. For more information you can also view [What is the difference between OpenRewrite and Moderne?](https://docs.openrewrite.org/reference/faq#what-is-the-difference-between-openrewrite-and-moderne).

**Reply by kjy5** [reply]
Thanks for your reply! I'm interested in using OpenRewrite to make an AST (or in this case an LST) merge driver for java source code. Specifically, given some a three-way merge scenario, I'd like to use OpenRewrite's visitor in a recipe to create XML representations of the source code, pass it to some XML merging algorithm to merge them, then reconstruct java source code from the merged XML.

The intended result would be a standalone application that I can pass arbitrary three-way merge files into.

Let me know if this is possible with OpenRewrite!

**Reply by timtebeek** [reply]
Hmm; while I could understand the appeal to use our lossless parsers I feel it might not be a good fit to use OpenRewrite here. We've put a lot of effort into doing correct type attribution, which you'd not use much of if you're doing single file comparison, and would get in the way with serialization when types are missing, malformed, partial, or self referential to name a few.

I'm not saying the above issues can't be solved, but you'd effectively be on your own as LST serialization is a differentiator for Moderne. As much as I'd love to help and see you do well, in this case I can't help you very much. Hope you'll understand, as it's at times a tough balance between OSS and a business on top of OSS.

**Reply by kjy5** [reply]
Thanks again for your reply and breaking down the issues here with parsing.

Aside from the parsing issue, is there also no way to run the visitors in in a standalone application? Do I have to use them in a recipe in an existing project?

**Reply by timtebeek** [reply]
It's not really been set up to run isolated visitors against a single source file well; there's a lot of assumptions as to recipe life cycles, visitors and cursors, some of [which you can see here](https://github.com/openrewrite/rewrite/blob/fe15bdba4aca5bfe5fdc2c6f7b50ff10c90f6a98/rewrite-core/src/main/java/org/openrewrite/scheduling/RecipeRunCycle.java#L157-L167) , that might not translate well to your use case.

**Reply by kjy5** [reply]
All right. Thank you for your help!

**Reply by timtebeek** [reply]
You're welcome & I wish you well with the merge driver! :)
